autoconnect/autoconnect-ws/autoconnect-ws-sm/src/identified/on_client_msg.rs

```diff
@@ -0,0 +1,204 @@
+use std::collections::HashMap;
+
+use cadence::CountedExt;
+use uuid::Uuid;
+
+use autoconnect_common::protocol::{ClientAck, ClientMessage, ServerMessage};
+use autopush_common::{endpoint::make_endpoint, util::sec_since_epoch};
+
+use super::WebPushClient;
+use crate::error::SMError;
+
+impl WebPushClient {
+    pub async fn on_client_msg(
+        &mut self,
+        msg: ClientMessage,
+    ) -> Result<Vec<ServerMessage>, SMError> {
+        match msg {
+            ClientMessage::Hello { .. } => {
+                Err(SMError::InvalidMessage("Already Hello'd".to_owned()))
+            }
+            ClientMessage::Register { channel_id, key } => {
+                Ok(vec![self.register(channel_id, key).await?])
+            }
+            ClientMessage::Unregister { channel_id, code } => {
+                Ok(vec![self.unregister(channel_id, code).await?])
+            }
+            ClientMessage::BroadcastSubscribe { broadcasts } => Ok(self
+                .broadcast_subscribe(broadcasts)
+                .await?
+                .map_or_else(Vec::new, |smsg| vec![smsg])),
+            ClientMessage::Ack { updates } => self.ack(&updates).await,
+            ClientMessage::Nack { code, .. } => {
+                self.nack(code).await?;
+                Ok(vec![])
+            }
+            ClientMessage::Ping => Ok(vec![self.ping().await?]),
+        }
+    }
+
+    async fn register(
+        &mut self,
+        channel_id_str: String,
+        key: Option<String>,
+    ) -> Result<ServerMessage, SMError> {
+        trace!("WebPushClient:register";
+               "uaid" => &self.uaid.to_string(),
+               "channel_id" => &channel_id_str,
+               "key" => &key,
+        );
+        let channel_id = Uuid::try_parse(&channel_id_str)
+            .map_err(|_| SMError::InvalidMessage(format!("Invalid channelID: {channel_id_str}")))?;
+        if channel_id.as_hyphenated().to_string() != channel_id_str {
+            return Err(SMError::InvalidMessage(format!(
+                "Invalid UUID format, not lower-case/dashed: {channel_id}",
+            )));
+        }
+
+        let smsg = match self._register(&channel_id, key).await {
+            Ok(endpoint) => {
+                let _ = self.app_state.metrics.incr("ua.command.register");
+                self.stats.registers += 1;
+                ServerMessage::Register {
+                    channel_id,
+                    status: 200,
+                    push_endpoint: endpoint,
+                }
+            }
+            Err(e) => {
+                error!("WebPushClient::register failed: {}", e);
+                ServerMessage::Register {
+                    channel_id,
+                    status: 500,
+                    push_endpoint: "".to_owned(),
+                }
+            }
+        };
+        Ok(smsg)
+    }
+
+    async fn _register(
```

Comment:
```suggestion
    async fn get_registration_endpoint(
```

Not really a fan of the _foo funcitons. Not a hill I'm going to die on, but I don't find them more readable.

Reply:
I'll go with `do_` instead which I've used in at least one other place (syncstorage).